File: src/scenes/play/Spaceship.cpp
#include "Spaceship.hpp"
#include "Bullet.hpp"
#include <CppInvaders.hpp>

#define X_START 16
#define X_BORDER 24
#define Y 216
#define VX 80

#define TIME_DEPLOYING 2.0f
#define TIME_DEAD 1.5f
#define TIME_EXPLOSION_FRAME 0.1f

Spaceship::Spaceship() {
    state = DEPLOYING;
    x = X_START;
    timer = 0;
    lives = 3;
}

bool Spaceship::collide_rect(Pico_Abs_Rect rct, Pico_Anchor anc) const {
    Pico_Rel_Rect ship_rct = { '!', {(float)SDL_roundf(x), (float)Y, 15, 8}, PICO_ANCHOR_N, NULL };
    Pico_Rel_Rect other_rct = { '!', {(float)rct.x, (float)rct.y, (float)rct.w, (float)rct.h}, anc, NULL };
    return state == DEPLOYED && pico_vs_rect_rect(&ship_rct, &other_rct);
}

void Spaceship::explode() {
    pico_output_sound(SFX_SPACESHIP_KILLED);
    state = EXPLODING;
    timer = 0;
    explosion_frames = 0;
}

Bullet* Spaceship::shoot() {
    pico_output_sound(SFX_SPACESHIP_SHOOT);
    return new Bullet(x, Y, -200);
}

void Spaceship::update(float delta) {
    const Uint8 *keys = SDL_GetKeyboardState(nullptr);
    Pico_Abs_Dim size;
    pico_get_view(NULL, &size, NULL, NULL, NULL, NULL, NULL, NULL);
    timer += delta;

    switch (state) {
    case DEPLOYING:
        if (timer >= TIME_DEPLOYING) {
            state = DEPLOYED;
            x = X_BORDER;
            timer -= TIME_DEPLOYING;
        }
        break;
    case DEPLOYED:
        x -= VX * delta * (keys[SDL_SCANCODE_A] || keys[SDL_SCANCODE_LEFT]);
        x += VX * delta * (keys[SDL_SCANCODE_D] || keys[SDL_SCANCODE_RIGHT]);
        x = SDL_max(X_BORDER, SDL_min(size.w - X_BORDER, x));
        timer -= delta;
        break;
    case EXPLODING:
        if (timer >= TIME_EXPLOSION_FRAME) {
            timer -= TIME_EXPLOSION_FRAME;
            if (++explosion_frames >= 10) {
                state = DEPLOYING;
                lives--;
            }
        }
        break;
    }
}

void Spaceship::draw() const {
    if (state == DEPLOYING) return;

    float draw_x = (float)SDL_roundf(x);
    float draw_y = (float)Y;
    
    char key[256];
    Pico_Rel_Rect crop = { '!', {0, 0, 16, 8}, PICO_ANCHOR_NW, NULL };

    if (state == DEPLOYED) {
        snprintf(key, sizeof(key), "/crop/ship/deployed");
        crop.x = 0;
    } else if (state == EXPLODING) {
        snprintf(key, sizeof(key), "/crop/ship/exploding/%d", explosion_frames % 2);
        crop.x = 16.0f * (1 + explosion_frames % 2);
    }

    pico_layer_image_mode('=', "ship", IMG_SPACESHIP);
    pico_layer_sub_mode('=', key, "ship", &crop);
    Pico_Rel_Rect dst = { '!', {draw_x, draw_y, 16, 8}, PICO_ANCHOR_N, NULL };
    pico_output_draw_layer(key, &dst);
}

File: lua/scenes/play/Spaceship.lua
local class = require "class"
local assets = require "assets"
local Bullet = require "scenes.play.Bullet"

local Spaceship = class.new()

function Spaceship:init(game_state)
    self.state = "DEPLOYING"
    self.x = 16
    self.timer = 0
    self.lives = 3
    self.explosion_frames = 0
    self.game_state = game_state
end

function Spaceship:collide_rect(rct, anc)
    if self.state ~= "DEPLOYED" then return false end
    local ship_rct = { '!', x=math.floor(self.x + 0.5), y=216, w=15, h=8, anchor='N' }
    local other_rct = { '!', x=rct.x, y=rct.y, w=rct.w, h=rct.h, anchor=anc }
    return pico.vs.rect_rect(ship_rct, other_rct)
end

function Spaceship:explode()
    pico.output.sound(assets.SFX_SPACESHIP_KILLED)
    self.state = "EXPLODING"
    self.timer = 0
    self.explosion_frames = 0
end

function Spaceship:shoot()
    pico.output.sound(assets.SFX_SPACESHIP_SHOOT)
    return Bullet:new(self.x, 216, -200)
end

function Spaceship:update(delta)
    local view = pico.get.view()
    local size = view.dim
    self.timer = self.timer + delta

    if self.state == "DEPLOYING" then
        if self.timer >= 2.0 then
            self.state = "DEPLOYED"
            self.x = 24
            self.timer = self.timer - 2.0
        end
    elseif self.state == "DEPLOYED" then
        local left = (self.game_state.keys["A"] or self.game_state.keys["Left"]) and 1 or 0
        local right = (self.game_state.keys["D"] or self.game_state.keys["Right"]) and 1 or 0
        
        self.x = self.x - 80 * delta * left
        self.x = self.x + 80 * delta * right
        self.x = math.max(24, math.min(size.w - 24, self.x))
        self.timer = self.timer - delta
    elseif self.state == "EXPLODING" then
        if self.timer >= 0.1 then
            self.timer = self.timer - 0.1
            self.explosion_frames = self.explosion_frames + 1
            if self.explosion_frames >= 10 then
                self.state = "DEPLOYING"
                self.lives = self.lives - 1
            end
        end
    end
end

function Spaceship:draw()
    if self.state == "DEPLOYING" then return end

    local draw_x = math.floor(self.x + 0.5)
    local draw_y = 216
    local key = ""
    local crop = { '!', x=0, y=0, w=16, h=8, anchor='NW' }

    if self.state == "DEPLOYED" then
        key = "/crop/ship/deployed"
        crop.x = 0
    elseif self.state == "EXPLODING" then
        key = "/crop/ship/exploding/" .. (self.explosion_frames % 2)
        crop.x = 16.0 * (1 + (self.explosion_frames % 2))
    end

    pico.layer.image('=', "ship", assets.IMG_SPACESHIP)
    pico.layer.sub('=', key, "ship", crop)
    local dst = { '!', x=draw_x, y=draw_y, w=16, h=8, anchor='N' }
    pico.output.draw.layer(key, dst)
end

return Spaceship
